docs/design_notes/jax_versioning.md

```diff
@@ -0,0 +1,175 @@
+
+Each `jax` release has an associated minimum `jaxlib` version `mx.my.mz`. The
+minimum `jaxlib` version for `jax` version `x.y.z` must be no greater than
+`x.y.0`.
+
+For `jax` version `x.y.z` and `jaxlib` version `lx.ly.lz` to be compatible,
+the following must hold:
+
+* The jaxlib version (`lx.ly.lz`) must be greater than or equal to the minimum
+  jaxlib version (`mx.my.mz`).
+* The jax version (`x.y.z`) must have major and minor versions (i.e., `x.y`)
+  greater than or equal to the jaxlib major and minor versions (i.e., `mx.my`).
+  For example, `jax` 3.4.5 is compatible with `jaxlib` 3.4.9, but not `jaxlib`
+  3.5.0 or 4.1.3.
+
+These
+[version constraints](https://github.com/google/jax/blob/main/jax/version.py)
+are currently checked by `jax` at import time, instead of being expressed as
+Python package version constraints. `jax` checks the `jaxlib` version at
+runtime rather than using a `pip` package version constraint because we
+[provide separate `jaxlib` wheels](https://github.com/google/jax#installation)
+for a variety of hardware and software versions (e.g, GPU, TPU, etc.). Since we
+do not know which is the right choice for any given user, we do not want `pip`
+to install a `jaxlib` package for us automatically.
+
+In the future, we hope to separate out the hardware-specific pieces of `jaxlib`
+into separate plugins, at which point the minimum version could be expressed as
+a Python package dependency.
+
+## How can I safely make changes to the API of `jaxlib`?
+
+* Where possible, changes to `jaxlib` should be made in a backwards-compatible
+  way. At a minimum changes to `jaxlib` must maintain API compatibility with all
+  past and future `jax` releases that share the same major and minor version
+  numbers.
+
+  For example, it is usually safe to add a new function to `jaxlib`, but unsafe
+  to remove an existing function or to change its signature without potentially
+  breaking older `jax` versions. Changes to `jax` must work or fail gracefully
+  for all `jaxlib` releases greater than the minimum up to the current
+  `major.minor` version.
+
+* `jax` may drop compatibility with older `jaxlib`s at any time by increasing
+  the minimum jaxlib version. However, the minimum jaxlib version must be no
+  greater than the current `major.minor.0` version of `jax`, or the most recent
+  `jaxlib` release prior to `major.minor`, whichever is older.
+
+  In other words, it should always be sufficient to use `jaxlib` releases with
+  the same `major.minor` version as the `jax` release, if one exists, or the
+  most recent release before that, if not.
+
+  For example, to remove an old backwards compatibility path in the `jax` Python
+  code, it is sufficient to bump the minimum jaxlib version and then delete the
+  compatibility path.
+
+* For `jaxlib` to make a change that is not compatible with existing `jax`
+  releases, its `major.minor` version numbers must be incremented. The version
```

Comment:
If we make let's say 4 or 5 incompatible changes then we'll be at jaxlib 4.xx, or 5.xx. Is that desirable?

Reply:
I just pushed a new version of the PR, but note that *nothing* either past or previous requires that you bump the major version. Bumping the minor version would have been enough in the previous scheme. Now bumping *any* version is enough: you just have to release `jax` every time you release `jaxlib`.